// @openzeppelin/upgrades-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertNotProxy = exports.DeployBeaconProxyKindError = exports.DeployBeaconProxyImplUnknownError = exports.DeployBeaconProxyUnsupportedError = exports.PrepareUpgradeUnsupportedError = exports.LoadProxyUnsupportedError = exports.BeaconProxyUnsupportedError = exports.isBeacon = exports.getImplementationAddressFromProxy = exports.getImplementationAddressFromBeacon = exports.isBeaconProxy = exports.isTransparentOrUUPSProxy = exports.UpgradeableContract = exports.processProxyKind = exports.setProxyKind = exports.logWarning = exports.getStorageLayoutForAddress = void 0;
__exportStar(require("./validate"), exports);
__exportStar(require("./impl-store"), exports);
__exportStar(require("./version"), exports);
__exportStar(require("./manifest"), exports);
__exportStar(require("./storage"), exports);
__exportStar(require("./eip-1967"), exports);
__exportStar(require("./provider"), exports);
__exportStar(require("./src-decoder"), exports);
__exportStar(require("./solc-api"), exports);
__exportStar(require("./deployment"), exports);
__exportStar(require("./link-refs"), exports);
__exportStar(require("./error"), exports);
var manifest_storage_layout_1 = require("./manifest-storage-layout");
Object.defineProperty(exports, "getStorageLayoutForAddress", { enumerable: true, get: function () { return manifest_storage_layout_1.getStorageLayoutForAddress; } });
__exportStar(require("./scripts/migrate-oz-cli-project"), exports);
var log_1 = require("./utils/log");
Object.defineProperty(exports, "logWarning", { enumerable: true, get: function () { return log_1.logWarning; } });
var proxy_kind_1 = require("./proxy-kind");
Object.defineProperty(exports, "setProxyKind", { enumerable: true, get: function () { return proxy_kind_1.setProxyKind; } });
Object.defineProperty(exports, "processProxyKind", { enumerable: true, get: function () { return proxy_kind_1.processProxyKind; } });
var standalone_1 = require("./standalone");
Object.defineProperty(exports, "UpgradeableContract", { enumerable: true, get: function () { return standalone_1.UpgradeableContract; } });
var eip_1967_type_1 = require("./eip-1967-type");
Object.defineProperty(exports, "isTransparentOrUUPSProxy", { enumerable: true, get: function () { return eip_1967_type_1.isTransparentOrUUPSProxy; } });
Object.defineProperty(exports, "isBeaconProxy", { enumerable: true, get: function () { return eip_1967_type_1.isBeaconProxy; } });
var impl_address_1 = require("./impl-address");
Object.defineProperty(exports, "getImplementationAddressFromBeacon", { enumerable: true, get: function () { return impl_address_1.getImplementationAddressFromBeacon; } });
Object.defineProperty(exports, "getImplementationAddressFromProxy", { enumerable: true, get: function () { return impl_address_1.getImplementationAddressFromProxy; } });
var beacon_1 = require("./beacon");
Object.defineProperty(exports, "isBeacon", { enumerable: true, get: function () { return beacon_1.isBeacon; } });
var usage_error_1 = require("./usage-error");
Object.defineProperty(exports, "BeaconProxyUnsupportedError", { enumerable: true, get: function () { return usage_error_1.BeaconProxyUnsupportedError; } });
Object.defineProperty(exports, "LoadProxyUnsupportedError", { enumerable: true, get: function () { return usage_error_1.LoadProxyUnsupportedError; } });
Object.defineProperty(exports, "PrepareUpgradeUnsupportedError", { enumerable: true, get: function () { return usage_error_1.PrepareUpgradeUnsupportedError; } });
Object.defineProperty(exports, "DeployBeaconProxyUnsupportedError", { enumerable: true, get: function () { return usage_error_1.DeployBeaconProxyUnsupportedError; } });
Object.defineProperty(exports, "DeployBeaconProxyImplUnknownError", { enumerable: true, get: function () { return usage_error_1.DeployBeaconProxyImplUnknownError; } });
Object.defineProperty(exports, "DeployBeaconProxyKindError", { enumerable: true, get: function () { return usage_error_1.DeployBeaconProxyKindError; } });
Object.defineProperty(exports, "assertNotProxy", { enumerable: true, get: function () { return usage_error_1.assertNotProxy; } });
