// @hashgraph/sdk/lib/account/TokenNftAllowance.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _TokenId = _interopRequireDefault(require("../token/TokenId.cjs"));

var _AccountId = _interopRequireDefault(require("./AccountId.cjs"));

var _long = _interopRequireDefault(require("long"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").INftAllowance} proto.INftAllowance
 * @typedef {import("@hashgraph/proto").ITokenID} proto.ITokenID
 * @typedef {import("@hashgraph/proto").IAccountID} proto.IAccountID
 */
class TokenNftAllowance {
  /**
   * @internal
   * @param {object} props
   * @param {TokenId} props.tokenId
   * @param {AccountId} props.spenderAccountId
   * @param {Long[] | null} props.serialNumbers
   */
  constructor(props) {
    /**
     * The token that the allowance pertains to.
     *
     * @readonly
     */
    this.tokenId = props.tokenId;
    /**
     * The account ID of the spender of the hbar allowance.
     *
     * @readonly
     */

    this.spenderAccountId = props.spenderAccountId;
    /**
     * The current balance of the spender's token allowance.
     * **NOTE**: If `null`, the spender has access to all of the account owner's NFT instances
     * (currently owned and any in the future).
     *
     * @readonly
     */

    this.serialNumbers = props.serialNumbers;
    Object.freeze(this);
  }
  /**
   * @internal
   * @param {proto.INftAllowance} allowance
   * @returns {TokenNftAllowance}
   */


  static _fromProtobuf(allowance) {
    return new TokenNftAllowance({
      tokenId: _TokenId.default._fromProtobuf(
      /** @type {proto.ITokenID} */
      allowance.tokenId),
      spenderAccountId: _AccountId.default._fromProtobuf(
      /** @type {proto.IAccountID} */
      allowance.spender),
      serialNumbers: allowance.approvedForAll != null && allowance.approvedForAll.value ? null : allowance.serialNumbers != null ? allowance.serialNumbers.map(serialNumber => _long.default.fromValue(serialNumber)) : []
    });
  }
  /**
   * @internal
   * @returns {proto.INftAllowance}
   */


  _toProtobuf() {
    return {
      tokenId: this.tokenId._toProtobuf(),
      spender: this.spenderAccountId._toProtobuf(),
      approvedForAll: this.serialNumbers == null ? {
        value: true
      } : null,
      serialNumbers: this.serialNumbers
    };
  }

}

exports.default = TokenNftAllowance;
