// defender-base-client/lib/utils/network.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toChainId = exports.fromChainId = exports.isValidNetwork = exports.Networks = void 0;
const lodash_1 = require("lodash");
exports.Networks = [
    'mainnet',
    'rinkeby',
    'ropsten',
    'kovan',
    'goerli',
    'xdai',
    'sokol',
    'fuse',
    'bsc',
    'bsctest',
    'fantom',
    'fantomtest',
    'moonbase',
    'matic',
    'mumbai',
];
function isValidNetwork(text) {
    return exports.Networks.includes(text);
}
exports.isValidNetwork = isValidNetwork;
function fromChainId(chainId) {
    return lodash_1.findKey(chainIds, (number) => number === chainId);
}
exports.fromChainId = fromChainId;
function toChainId(network) {
    return chainIds[network];
}
exports.toChainId = toChainId;
const chainIds = {
    mainnet: 1,
    ropsten: 3,
    rinkeby: 4,
    goerli: 5,
    kovan: 42,
    xdai: 100,
    sokol: 77,
    fuse: 122,
    bsc: 56,
    bsctest: 97,
    fantom: 250,
    fantomtest: 0xfa2,
    moonbase: 1287,
    matic: 137,
    mumbai: 80001,
};
