// defender-base-client/lib/api/api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAuthenticatedApi = exports.createApi = void 0;
const axios_1 = __importDefault(require("axios"));
const lodash_1 = require("lodash");
const auth_1 = require("./auth");
function createApi(key, token, apiUrl) {
    const instance = axios_1.default.create({
        baseURL: apiUrl,
        headers: {
            'X-Api-Key': key,
            Authorization: `Bearer ${token}`,
            'Content-Type': 'application/json',
        },
    });
    instance.interceptors.response.use((response) => response.data, (error) => Promise.reject({
        response: lodash_1.pick(error.response, 'status', 'statusText', 'data'),
        message: error.message,
        request: lodash_1.pick(error.request, 'path', 'method'),
    }));
    return instance;
}
exports.createApi = createApi;
async function createAuthenticatedApi(userPass, poolData, apiUrl) {
    const token = await auth_1.authenticate(userPass, poolData);
    const api = createApi(userPass.Username, token, apiUrl);
    return api;
}
exports.createAuthenticatedApi = createAuthenticatedApi;
